### PR TITLE
Getcodedatasets

### PR DESCRIPTION
**Pull request by peterhoward42**
### What

Real (not temp mock) implementation of codelist.GetCodeDatasets method

### How to review

In addition to the usual code scrutiny...

Check the go generate step to make the db moq works as follows:

            cd dp-graph/neptune/driver
            go generate

Run unit tests in package

            cd dp-graph/neptune
            go test ./...

Build code list api with this dp-graph, run it (configured -> Neptune), and hit this endpoint...

           http://localhost:22400/code-lists/code-list-id-for-dataset-test/editions/code-list-edition-for-dataset-test/codes/code-value-for-dataset-test/datasets

Check for these results:

- item.count field should be 1 (at bottom)
- item.links.self.id should be "code-value-for-dataset-test"
- items.dimension_label should be "test-dimension-name"
- should contain editions sub-block showing:
- 10 editions with suffix 0...9
- each showing latest v of 9

### Who can review

Describe who worked on the changes  (Pete), so that other people can review. @eldeal

### CODE REVIEW

File: neptune/codelistsdataset.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/*
+buildDim2Edition consumes a list of triples such as
+["dimName1", "datasetEdition1", "version4"], and builds a dim2Edition
+structure based on the latest versions available for each combination of
+dimension name and dataset edition.
+*/
+func buildDim2Edition(responseTriples [][]string) (dim2Edition, error) {
+	d2e := dim2Edition{}
+	for _, respTriple := range responseTriples {
+		dimensionName := respTriple[0]
```
Comment (eldeal) [inline neptune/codelistsdataset.go]: (lost this comment in the git-wilderness, so apologies if this ends up as a duplicate)

Unfortunately, nesting under dimensionName won't be specific enough for our case, we need to be including the Dataset ID as a differentiator.

https://api.beta.ons.gov.uk/v1/code-lists/admin-geography/editions/one-off/codes/K02000001/datasets
If you consider this endpoint, the dimesnion Name is the same across all of the subdocuments, and their editions labels are largely the same as well, but there is a separate document for every dataset ID.

We'll need to include projecting the dataset_ID field in the query, and expand these helpers to also separate results based on it.

File: neptune/codelistsdataset.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+createTriples splits a list of strings into clumps of 3
+*/
+func createTriples(responses []string) ([][]string, error) {
+	var responseTriples = [][]string{}
+	const stride = 3 // I.e. dimesionName, edition, version
+	if len(responses)%stride != 0 {
+		return nil, errors.New("List length is not divisible by 3")
+	}
+	nInstances := len(responses) / stride
+	for i := 0; i < nInstances; i++ {
```
[the remark this answers is not in the document]
Comment (peterhoward42) [reply neptune/codelistsdataset.go]: fixed and pushed

File: neptune/codelistsdataset.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		offset := i * stride
+		dimensionName := responses[offset+0]
+		dataSetEdition := responses[offset+1]
```
[the remark this answers is not in the document]
Comment (peterhoward42) [reply neptune/codelistsdataset.go]: fixed and pushed.

File: neptune/codelistsdataset.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		TotalCount: len(d2e),
+	}
+	for dimensionName, e2v := range d2e {
+		datasetLinks := &models.DatasetLinks{Self: &models.Link{ID: code}}
+		dataset := models.Dataset{
+			Links:          datasetLinks,
+			DimensionLabel: dimensionName,
+			Editions:       []models.DatasetEdition{},
+		}
+		for dataSetEdition, version := range e2v {
```
[the remark this answers is not in the document]
Comment (peterhoward42) [reply neptune/codelistsdataset.go]: fixed and pushed.

File: neptune/codelistsdataset.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+*/
+func buildResponse(d2e dim2Edition, code string, codeListID string) *models.Datasets {
+	datasets := &models.Datasets{
+		Items:      []models.Dataset{},
+		Count:      len(d2e),
+		Limit:      len(d2e),
+		TotalCount: len(d2e),
+	}
+	for dimensionName, e2v := range d2e {
+		datasetLinks := &models.DatasetLinks{Self: &models.Link{ID: code}}
```
[the remark this answers is not in the document]
Comment (peterhoward42) [reply neptune/codelistsdataset.go]: fixed and pushed.

File: neptune/codelistsdataset.go (hunk 1 above)
Comment (peterhoward42) [reply neptune/codelistsdataset.go]: fixed and pushed.

**Pull request merged**